presto-main/src/main/java/com/facebook/presto/operator/scalar/DataSizeFunctions.java

```diff
@@ -0,0 +1,136 @@
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlType;
+import io.airlift.slice.Slice;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.StandardErrorCode.NUMERIC_VALUE_OUT_OF_RANGE;
+import static com.facebook.presto.spi.type.Decimals.encodeUnscaledValue;
+import static java.lang.Character.isDigit;
+import static java.lang.String.format;
+
+public class DataSizeFunctions
+{
+    private DataSizeFunctions() {}
+
+    @Description("converts data size string to bytes")
+    @ScalarFunction("parse_data_size")
+    @LiteralParameters("x")
+    @SqlType("decimal(38,0)")
+    public static Slice parseDataSize(@SqlType("varchar(x)") Slice input)
+    {
+        String dataSize = input.toStringUtf8();
+
+        int valueLength = 0;
+        for (int i = 0; i < dataSize.length(); i++) {
+            char c = dataSize.charAt(i);
+            if (isDigit(c) || c == '.') {
+                valueLength++;
+            }
+            else {
+                break;
+            }
+        }
+
+        checkDataSize(valueLength > 0, dataSize);
+        String valueString = dataSize.substring(0, valueLength);
+        BigDecimal value = parseValue(valueString, dataSize);
+        String unitString = dataSize.substring(valueLength);
+        Unit unit = parseUnit(unitString, dataSize);
+
+        BigInteger bytes = value.multiply(unit.getFactor()).toBigInteger();
+        try {
+            return encodeUnscaledValue(bytes);
+        }
+        catch (ArithmeticException e) {
+            throw new PrestoException(NUMERIC_VALUE_OUT_OF_RANGE, format("Value out of range: '%s' ('%sB')", dataSize, bytes));
+        }
+    }
+
+    private static BigDecimal parseValue(String value, String dataSize)
+    {
+        try {
+            return new BigDecimal(value);
+        }
+        catch (NumberFormatException e) {
+            throw invalidDataSize(dataSize);
+        }
+    }
+
+    private static Unit parseUnit(String unitString, String dataSize)
+    {
+        if (unitString.isEmpty()) {
+            return Unit.BYTE;
+        }
+        for (Unit unit : Unit.values()) {
+            if (unit.getUnitString().equalsIgnoreCase(unitString)) {
+                return unit;
+            }
+        }
+        throw invalidDataSize(dataSize);
+    }
+
+    private static void checkDataSize(boolean condition, String dataSize)
+    {
+        if (!condition) {
+            throw invalidDataSize(dataSize);
+        }
+    }
+
+    private static PrestoException invalidDataSize(String dataSize)
+    {
+        return new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Invalid data size: '%s'", dataSize));
+    }
+
+    private enum Unit
+    {
+        BYTE(BigDecimal.ONE, "B"),
+        KILOBYTE(new BigDecimal(1L << 10), "KB"),
```

Comment:
We use `kB` in `DataSize`. Using uppercase here make this function incompatible with that (unless you uppercase before calling it). I believe it uses lowercase `k` since that is how the SI unit is written.

Reply:
I'm doing `equalsIgnoreCase` when getting the unit.